### ISSUE TITLE
ENH: np.asarray should have an option to ensure the array returned is writeable

### ISSUE THREAD

**Issue by bashtage** [numpy/numpy#23625]
### Proposed new feature or change:

Pandas is moving toward copy-on-write.  When CoW is enabled, calling `np.asarray(df_with_cow)` returns a read-only version of an ndarray.  This breaks a lot of existing code.  This probably isn't precisely a numpy issue, since `asarray` is just doing what is promises.

Perhaps `np.asarray(x, writable=True)` so that  copy would be taken from anything that is read only.

As a small measure of this issue, enabling CoW leads to about 300 failures on statsmodels tests.

**Comment by seberg** [comment]
`np.require` is a way to spell this (not really better than `if not a.flags.writeable: a = a.copy("K")`).

Adding a `kwarg` to the `(as)?array*` family of functions seems reasonable to me to make it easier (I also just don't like the `require` name that much, I suspect).

**Comment by rgommers** [comment]
Before concluding something here, let's see what the suggested usage pattern is here? On the linked pandas issue the first reply is "this was intended", so adding a flag to `asarray` to make it easy to overrule that intent may not be quite right?

**Comment by bashtage** [comment]
Some common use patterns is 

```
def impute_missing_values(x: ArrayLike):
    y = np.array(x)
    missing = np.isnan(x)
    y[missing] = some_fill_value
    if is_pandas(x):
        x.loc[missing] = y
        return x
    return y
```

Another one that is cause lots of problem is in Cython

```
def f(x):
    cdef double[::1] y
    y = np.ascontiguousarray(x)
```

which raises since y is now read only which x is a standard pandas Series.  

The around for now is to liberally use `requires` with `requirements="W"` in place of `asarray`, but the move to CoW is looking to be painful since it is changing how numpy and pandas interact.

**Comment by seberg** [comment]
I agree with Ralf.  Pandas seems to do the right thing when exporting it as readonly, and we to narrow down the statsmodel/downstream use-case here to see what is useful to help.

If you do `np.array()` you should get a writeable copy.  And for most cython code `cdef const double[::1]` would hopefully be the best solution anyway (to accept readonly inputs).

(If `np.array()` returns a non-writeable object than would seem like a bug somewhere that we need to track down!)

There is some logic in the first example that looks like "operate in-place if-possible" and that part breaks for pandas (but is not really used by the pandas branch anyway?).  So I also wonder if the solution isn't to either always copy with `np.array()` or branch on `is_pandas`?

I guess it may be tedious to figure out where the copy logic needs to go in practice, since you may already have copied the pandas dataframe thinking that this should be enough.

**Comment by bashtage** [comment]
I think the question, which might be for pandas, is whether `asarray` should give a read-only view to the same data, or should return a copy.  If `Series` and `DataFrames` where sufficiently foreign to numpy, then you would just get a copy that would be writeable.  Instead you get a fragile view of the data, which feels like it is going to be very hard to know is right.

**Comment by rgommers** [comment]
> I think the question, which might be for pandas, is whether `asarray` should give a read-only view to the same data, or should return a copy.

I think it should give a hard error, always. The transition path to get there is a bit up to pandas of course, and may take time. But planning that would be useful.

**Comment by rgommers** [comment]
@jorisvandenbossche is there a plan already for how to deal with numpy-pandas interop? This kind of question is going to happen a lot more when both nullable dtypes and copy-on-write get used more.
